dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp

```diff
@@ -163,6 +200,328 @@ PrehandleResult KVStore::preHandleSnapshotToFiles(
     return result;
 }
 
+static inline std::vector<std::string> getSplitKey(
+    LoggerPtr log,
+    KVStore * kvstore,
+    RegionPtr new_region,
+    std::shared_ptr<DM::SSTFilesToBlockInputStream> sst_stream)
+{
+    // We don't use this is the single snapshot is small, due to overhead in decoding.
+    // TODO(split) find solution if the snapshot has too many untrimmed data.
+    // TODO(split) recover this after integration test passes.
+    // constexpr size_t PARALLEL_PREHANDLE_THRESHOLD = 1 * 1024 * 1024 * 1024;
+    constexpr size_t PARALLEL_PREHANDLE_THRESHOLD = 1 * 1024 * 1024;
+    size_t parallel_prehandle_threshold = PARALLEL_PREHANDLE_THRESHOLD;
+    fiu_do_on(FailPoints::force_set_parallel_prehandle_threshold, {
+        if (auto v = FailPointHelper::getFailPointVal(FailPoints::force_set_parallel_prehandle_threshold); v)
+            parallel_prehandle_threshold = std::any_cast<size_t>(v.value());
+    });
+    // If size is 0, do no parallel prehandle for this snapshot, which is legacy.
+    // If size is non-zero, use extra this many threads to prehandle.
+    std::vector<std::string> split_keys;
+    // Don't change the order of following checks, `getApproxBytes` involves some overhead,
+    // although it is optimized to bring about the minimum overhead.
+
+    if (new_region->getClusterRaftstoreVer() == RaftstoreVer::V2 && kvstore->getOngoingPrehandleTaskCount() < 2
```

Comment:
```
    ... && kvstore->getOngoingPrehandleTaskCount() < 2
```

So when TiFlash is running pre-handle for one large Region snapshot, then the coming concurrent Region snapshot decoding will receive empty `split_keys` and run in single thread?

Reply:
`kvstore->getOngoingPrehandleTaskCount() < 2` ensures the coming snapshot is the only prehanding snapshot.
If another snapshot comes later before the current prehandling finishes, `kvstore->getOngoingPrehandleTaskCount() < 2` is false, so it will be handle in single thread.
This is because we tend to use all available cores to prehandle when there is only one snapshot. It will bring about many context switches if we do this to another snapshot in the mean time.

Reply:
After discussion with @JaySon-Huang , this is a problem when there are a few but not a single big region snapshot. Meanwhile, if there are many cores, and a single big region, we will not use all cores to do prehandling, since it will result in too many splits.
These are two optimaztions, we will have another pr for this.